### ISSUE TITLE
When deleting and recreating a GCP node in an autoscaling group with bootstrapping on 1.7, pods sometimes not deleted by node controller

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#55091]
While testing bootstrapping in a 1.7 cluster, I noticed that deleting GCP instances in an autoscaling group (where the instance is deleted directly from the console) resulted in inconsistent behavior - sometimes pods on the deleted instance were properly deleted and rescheduled, other times they were not.

Scenario:

1. Create a 3 node autoscaling group, with nodes set to auto join the cluster
2. Delete a node instance in the GCP console (which results in the scaling group creating a new instance with the same name a minute or so later)
3. Observe that the node object always appears to be properly deleted (UID change) and the new node object is created extremely shortly thereafter by the new instance - 2-10s (bootstrap and auto approval are fast)
4. Observe that 75% of the time the pods on that node are NOT deleted, and also don't get their timestamps updated (they still have the status from the old node, including timestamps, but readiness is reset)
5. A few minutes later, the status is updated and shows up correctly for the new node (at least 1-2 minutes)

The issue is serious in this case because the networking plugin in use (openshift-sdn) correctly observes that a new node (with new UID) is created and assigns a new host subnet, but the kubelet simply starts the new containers.  Sometimes the pods are deleted which allows it to reset, and sometimes the plugin simply sees that the containers are restarted.

We definitely cannot guarantee that the kubelet knows the pod was scheduled for a different iteration of itself since the node name is the same.  The node controller is racing with the kubelet to observe the node deletion since the new node is created very close in time to the old node being deleted.  If the controller manager were shut down during this event, it would have no way to know that the node has changed.

At a minimum, the bug is that the status update lags until the next sync loop.  A larger error may be that we are inducing inconsistent behavior - the network plugin needs to follow along and react to the changing node, but it's not clear whether the pod should or should not be deleted.

```
# existing pod 
{
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": {
        "annotations": {
            "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"registry-console-1\",\"uid\":\"4b81a94b-c0eb-11e7-b34d-42010a8e0004\",\"apiVersion\":\"v1\",\"resourceVersion\":\"5599\"}}\n",
            "openshift.io/deployment-config.latest-version": "1",
            "openshift.io/deployment-config.name": "registry-console",
            "openshift.io/deployment.name": "registry-console-1",
            "openshift.io/generated-by": "OpenShiftNewApp",
            "openshift.io/scc": "restricted"
        },
        "creationTimestamp": "2017-11-03T23:42:58Z",
        "generateName": "registry-console-1-",
        "labels": {
            "app": "registry-console",
            "deployment": "registry-console-1",
            "deploymentconfig": "registry-console",
            "name": "registry-console"
        },
        "name": "registry-console-1-lc9g6",
        "namespace": "default",
        "ownerReferences": [
            {
                "apiVersion": "v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "ReplicationController",
                "name": "registry-console-1",
                "uid": "4b81a94b-c0eb-11e7-b34d-42010a8e0004"
            }
        ],
        "resourceVersion": "5683",
        "selfLink": "/api/v1/namespaces/default/pods/registry-console-1-lc9g6",
        "uid": "b8e9cdc0-c0f0-11e7-b34d-42010a8e0004"
    },
    "spec": {
        "containers": [
            {
                "env": [
                    {
                        "name": "OPENSHIFT_OAUTH_PROVIDER_URL",
                        "value": "https://api.claytontest.origin-ci-int-gce.dev.rhcloud.com/"
                    },
                    {
                        "name": "OPENSHIFT_OAUTH_CLIENT_ID",
                        "value": "cockpit-oauth-client"
                    },
                    {
                        "name": "KUBERNETES_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "COCKPIT_KUBE_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "REGISTRY_ONLY",
                        "value": "true"
                    },
                    {
                        "name": "REGISTRY_HOST",
                        "value": "docker-registry-default.apps.claytontest.origin-ci-int-gce.dev.rhcloud.com"
                    }
                ],
                "image": "cockpit/kubernetes:latest",
                "imagePullPolicy": "Always",
                "livenessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "initialDelaySeconds": 10,
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "name": "registry-console",
                "ports": [
                    {
                        "containerPort": 9090,
                        "protocol": "TCP"
                    }
                ],
                "readinessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "resources": {},
                "securityContext": {
                    "capabilities": {
                        "drop": [
                            "KILL",
                            "MKNOD",
                            "SETGID",
                            "SETUID"
                        ]
                    },
                    "privileged": false,
                    "runAsUser": 1000000000,
                    "seLinuxOptions": {
                        "level": "s0:c1,c0"
                    }
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File",
                "volumeMounts": [
                    {
                        "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                        "name": "default-token-r7r5f",
                        "readOnly": true
                    }
                ]
            }
        ],
        "dnsPolicy": "ClusterFirst",
        "imagePullSecrets": [
            {
                "name": "default-dockercfg-5czxp"
            }
        ],
        "nodeName": "ci-claytontest-ig-n-63c4",
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {
            "fsGroup": 1000000000,
            "seLinuxOptions": {
                "level": "s0:c1,c0"
            }
        },
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "volumes": [
            {
                "name": "default-token-r7r5f",
                "secret": {
                    "defaultMode": 420,
                    "secretName": "default-token-r7r5f"
                }
            }
        ]
    },
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:43:05Z",
                "status": "True",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
        "containerStatuses": [
            {
                "containerID": "docker://ef5aef49b57a811f642bba553a6a1c4dde865289c825234527e1c2b3aeceea12",
                "image": "docker.io/cockpit/kubernetes:latest",
                "imageID": "docker-pullable://docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256",
                "lastState": {},
                "name": "registry-console",
                "ready": true,
                "restartCount": 0,
                "state": {
                    "running": {
                        "startedAt": "2017-11-03T23:43:00Z"
                    }
                }
            }
        ],
        "hostIP": "10.142.0.2",
        "phase": "Running",
        "podIP": "172.16.0.4",
        "qosClass": "BestEffort",
        "startTime": "2017-11-03T23:42:58Z"
    }
}
```

Node is deleted, next update is only readiness change, not other container status

```
# pod UPDATE event - readiness changes, but not status info like transition time
{
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": {
        "annotations": {
            "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"registry-console-1\",\"uid\":\"4b81a94b-c0eb-11e7-b34d-42010a8e0004\",\"apiVersion\":\"v1\",\"resourceVersion\":\"5599\"}}\n",
            "openshift.io/deployment-config.latest-version": "1",
            "openshift.io/deployment-config.name": "registry-console",
            "openshift.io/deployment.name": "registry-console-1",
            "openshift.io/generated-by": "OpenShiftNewApp",
            "openshift.io/scc": "restricted"
        },
        "creationTimestamp": "2017-11-03T23:42:58Z",
        "generateName": "registry-console-1-",
        "labels": {
            "app": "registry-console",
            "deployment": "registry-console-1",
            "deploymentconfig": "registry-console",
            "name": "registry-console"
        },
        "name": "registry-console-1-lc9g6",
        "namespace": "default",
        "ownerReferences": [
            {
                "apiVersion": "v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "ReplicationController",
                "name": "registry-console-1",
                "uid": "4b81a94b-c0eb-11e7-b34d-42010a8e0004"
            }
        ],
        "resourceVersion": "7000",
        "selfLink": "/api/v1/namespaces/default/pods/registry-console-1-lc9g6",
        "uid": "b8e9cdc0-c0f0-11e7-b34d-42010a8e0004"
    },
    "spec": {
        "containers": [
            {
                "env": [
                    {
                        "name": "OPENSHIFT_OAUTH_PROVIDER_URL",
                        "value": "https://api.claytontest.origin-ci-int-gce.dev.rhcloud.com/"
                    },
                    {
                        "name": "OPENSHIFT_OAUTH_CLIENT_ID",
                        "value": "cockpit-oauth-client"
                    },
                    {
                        "name": "KUBERNETES_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "COCKPIT_KUBE_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "REGISTRY_ONLY",
                        "value": "true"
                    },
                    {
                        "name": "REGISTRY_HOST",
                        "value": "docker-registry-default.apps.claytontest.origin-ci-int-gce.dev.rhcloud.com"
                    }
                ],
                "image": "cockpit/kubernetes:latest",
                "imagePullPolicy": "Always",
                "livenessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "initialDelaySeconds": 10,
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "name": "registry-console",
                "ports": [
                    {
                        "containerPort": 9090,
                        "protocol": "TCP"
                    }
                ],
                "readinessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "resources": {},
                "securityContext": {
                    "capabilities": {
                        "drop": [
                            "KILL",
                            "MKNOD",
                            "SETGID",
                            "SETUID"
                        ]
                    },
                    "privileged": false,
                    "runAsUser": 1000000000,
                    "seLinuxOptions": {
                        "level": "s0:c1,c0"
                    }
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File",
                "volumeMounts": [
                    {
                        "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                        "name": "default-token-r7r5f",
                        "readOnly": true
                    }
                ]
            }
        ],
        "dnsPolicy": "ClusterFirst",
        "imagePullSecrets": [
            {
                "name": "default-dockercfg-5czxp"
            }
        ],
        "nodeName": "ci-claytontest-ig-n-63c4",
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {
            "fsGroup": 1000000000,
            "seLinuxOptions": {
                "level": "s0:c1,c0"
            }
        },
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "volumes": [
            {
                "name": "default-token-r7r5f",
                "secret": {
                    "defaultMode": 420,
                    "secretName": "default-token-r7r5f"
                }
            }
        ]
    },
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:43:05Z",
                "status": "False",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
        "containerStatuses": [
            {
                "containerID": "docker://ef5aef49b57a811f642bba553a6a1c4dde865289c825234527e1c2b3aeceea12",
                "image": "docker.io/cockpit/kubernetes:latest",
                "imageID": "docker-pullable://docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256",
                "lastState": {},
                "name": "registry-console",
                "ready": true,
                "restartCount": 0,
                "state": {
                    "running": {
                        "startedAt": "2017-11-03T23:43:00Z"
                    }
                }
            }
        ],
        "hostIP": "10.142.0.2",
        "phase": "Running",
        "podIP": "172.16.0.4",
        "qosClass": "BestEffort",
        "startTime": "2017-11-03T23:42:58Z"
    }
}
```

Eventually the status is updated to be accurate:

```
{
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": {
        "annotations": {
            "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"registry-console-1\",\"uid\":\"4b81a94b-c0eb-11e7-b34d-42010a8e0004\",\"apiVersion\":\"v1\",\"resourceVersion\":\"5599\"}}\n",
            "openshift.io/deployment-config.latest-version": "1",
            "openshift.io/deployment-config.name": "registry-console",
            "openshift.io/deployment.name": "registry-console-1",
            "openshift.io/generated-by": "OpenShiftNewApp",
            "openshift.io/scc": "restricted"
        },
        "creationTimestamp": "2017-11-03T23:42:58Z",
        "generateName": "registry-console-1-",
        "labels": {
            "app": "registry-console",
            "deployment": "registry-console-1",
            "deploymentconfig": "registry-console",
            "name": "registry-console"
        },
        "name": "registry-console-1-lc9g6",
        "namespace": "default",
        "ownerReferences": [
            {
                "apiVersion": "v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "ReplicationController",
                "name": "registry-console-1",
                "uid": "4b81a94b-c0eb-11e7-b34d-42010a8e0004"
            }
        ],
        "resourceVersion": "7244",
        "selfLink": "/api/v1/namespaces/default/pods/registry-console-1-lc9g6",
        "uid": "b8e9cdc0-c0f0-11e7-b34d-42010a8e0004"
    },
    "spec": {
        "containers": [
            {
                "env": [
                    {
                        "name": "OPENSHIFT_OAUTH_PROVIDER_URL",
                        "value": "https://api.claytontest.origin-ci-int-gce.dev.rhcloud.com/"
                    },
                    {
                        "name": "OPENSHIFT_OAUTH_CLIENT_ID",
                        "value": "cockpit-oauth-client"
                    },
                    {
                        "name": "KUBERNETES_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "COCKPIT_KUBE_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "REGISTRY_ONLY",
                        "value": "true"
                    },
                    {
                        "name": "REGISTRY_HOST",
                        "value": "docker-registry-default.apps.claytontest.origin-ci-int-gce.dev.rhcloud.com"
                    }
                ],
                "image": "cockpit/kubernetes:latest",
                "imagePullPolicy": "Always",
                "livenessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "initialDelaySeconds": 10,
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "name": "registry-console",
                "ports": [
                    {
                        "containerPort": 9090,
                        "protocol": "TCP"
                    }
                ],
                "readinessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "resources": {},
                "securityContext": {
                    "capabilities": {
                        "drop": [
                            "KILL",
                            "MKNOD",
                            "SETGID",
                            "SETUID"
                        ]
                    },
                    "privileged": false,
                    "runAsUser": 1000000000,
                    "seLinuxOptions": {
                        "level": "s0:c1,c0"
                    }
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File",
                "volumeMounts": [
                    {
                        "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                        "name": "default-token-r7r5f",
                        "readOnly": true
                    }
                ]
            }
        ],
        "dnsPolicy": "ClusterFirst",
        "imagePullSecrets": [
            {
                "name": "default-dockercfg-5czxp"
            }
        ],
        "nodeName": "ci-claytontest-ig-n-63c4",
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {
            "fsGroup": 1000000000,
            "seLinuxOptions": {
                "level": "s0:c1,c0"
            }
        },
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "volumes": [
            {
                "name": "default-token-r7r5f",
                "secret": {
                    "defaultMode": 420,
                    "secretName": "default-token-r7r5f"
                }
            }
        ]
    },
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:57:04Z",
                "status": "True",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
        "containerStatuses": [
            {
                "containerID": "docker://6c1218ff7a8d7e1cc462765c7e19e20cd8e69797dd27f20ffe12fb9a708e5d48",
                "image": "docker.io/cockpit/kubernetes:latest",
                "imageID": "docker-pullable://docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256",
                "lastState": {},
                "name": "registry-console",
                "ready": true,
                "restartCount": 0,
                "state": {
                    "running": {
                        "startedAt": "2017-11-03T23:57:00Z"
                    }
                }
            }
        ],
        "hostIP": "10.142.0.2",
        "phase": "Running",
        "podIP": "172.16.0.2",
        "qosClass": "BestEffort",
        "startTime": "2017-11-03T23:42:58Z"
    }
}
```

**Comment by smarterclayton** [comment]
@kubernetes/sig-node-bugs @sjenning @danwinship this is the root cause of openshift/origin#17125.  I don't think GCP allows you to assign new instance names in an autoscaling group (looks like it preserves to conserve names, can't find a flag from a cursory look).

**Comment by smarterclayton** [comment]
Didn't call it out above, but another bug is that a readiness status change is *not* updating lastTransitionTime from the node.  Not sure if that was fixed in 1.8 or 1.9
